### PR TITLE
feat: suggest fee (smart fee)

### PR DESCRIPTION
**Pull request by anxolin**
Implements the "smart" slippage calculations.

Follow up on https://github.com/cowprotocol/cow-sdk/pull/306

This order has been placed using the new SDK: https://explorer.cow.fi/sepolia/orders/0xb31414879f27655ab1e45b48f2e908700ae1d758172adf0aa9b2496f87870ff2016f34d4f2578c3e9dffc3f2b811ba30c0c9e7f36823a1f5?tab=overview

## Details
Mimics the logic used in CoW Swap with a few differences.

- The suggested BPS are rounded up instead of down.
- The new fee costs you get after applying the percentage, now rounds up (this is important, because we want to cover with the slippage at least the fee increase)
- Handle a few extra edge cases (see tests)
- It will never return BPS bigger than 10_000 --> 100%, or less than 50
   - CoW Swap allows only a maximum of 500 BPS. I've let it go all the way to 10_000. I think it should be UIs who recommend users to review the order. 
   - The only way to get to those numbers is if the fee is really significant compared to the sell volume, and I think in this case it makes sense to have a very high slippage. 
   - IMO It should be the responsibility of the UI to inform the user or add a cap. Yes, it can get the super high recommendation of slipage tolerance, but only if the order is super small.
   - Anyways, lets discuss the options, because here we can do a lot of fine-tuning

## Test
Adds exhaustive unit tests
Screenshot at May 13 18-56-56

Also, I deployed a RC: `6.0.0-RC.38`
This way, it should be easy to test this using NPM.

I used the NPM package and checked that:

The quote returns now a suggested slippage:
image

The app-data is updated (not the one from the quote, because we reuse the first quote we do with 50 BPS), but for posting the order w e
Screenshot at May 13 20-11-22

I posted an order using the SDK with auto-slippage:
image

https://explorer.cow.fi/sepolia/orders/0xb31414879f27655ab1e45b48f2e908700ae1d758172adf0aa9b2496f87870ff2016f34d4f2578c3e9dffc3f2b811ba30c0c9e7f36823a1f5?tab=overview

I also tested that if I provide a slippage (so not using auto-slippage), then I get to use the slippage I want (but still I get the suggested slippage in the result):
https://explorer.cow.fi/sepolia/orders/0x03cf60670ea839b2d16e39d8bf0b9c7c4098aa983e2ef1ce1bb164750c19b166016f34d4f2578c3e9dffc3f2b811ba30c0c9e7f36823a256?tab=overview

## Summary by CodeRabbit

- **New Features**
  - Introduced dynamic slippage suggestion based on order fees for more accurate trading recommendations.
  - Added new utilities for advanced percentage and basis point calculations.
- **Refactor**
  - Improved and modularized quote amount and cost calculations for better clarity and reusability.
- **Tests**
  - Added comprehensive tests for slippage suggestion logic to ensure reliability and correctness.

### CODE REVIEW

**Comment by anxolin** [comment]
@coderabbitai review

**Comment by anxolin** [comment]
@coderabbitai review

File: src/trading/suggestSlippageBpsFromFee.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
+import { applyPercentage, percentageToBps } from 'src/common/utils/math'
+
+const MIN_SLIPPAGE_BPS = 50
```
Comment (alfetopito) [inline src/trading/suggestSlippageBpsFromFee.ts:3]: If we don't have a cap of the max slippage (other than 100%), should we floor on the min slippage suggested?
For example, in case it's a large stable trade, does it still make sense to suggest 0.5%?

File: src/trading/suggestSlippageBpsFromFee.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+export interface SuggestSlippageBpsFromFeeParams {
+  /**
+   * Fee amount in the sell token returned by the quote
+   */
+  feeAmount: bigint
+
+  /**
+   * Actual amount to be sold
+   *
+   * For sell orders, its expected the sell amount after applying the fee is applied
```
Comment (alfetopito) [inline src/trading/suggestSlippageBpsFromFee.ts:18]:
```suggestion
   * For sell orders, its expected the sell amount after the fee is applied
```

File: src/trading/suggestSlippageBpsFromFee.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  // Multiplying factor must be a valid percentage
+  if (multiplyingFactorPercent < 0n) {
+    throw new Error('multiplyingFactorPercent must be a percentage: ' + multiplyingFactorPercent)
+  }
+
+  // Get the amount we want to account for our slippage
+  const feeAfterIncrease = applyPercentage(feeAmount, 100 + multiplyingFactorPercent)
+
+  // Account fee (fee is deducted to sell amount for sell orders and added to sell amount for buy orders)
+  const sellAmountAccountingFee = isSell ? sellAmount - feeAmount : sellAmount + feeAmount
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/trading/suggestSlippageBpsFromFee.ts:65]: If I understand the problem, in the original code I was considering the sell amount with fee fee already deduced, so when I multiply the fee amount by itself + 50% it's accounting for it twice.
Is that it?

File: src/trading/suggestSlippageBpsFromFee.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
+import { applyPercentage, percentageToBps } from 'src/common/utils/math'
+
+const MIN_SLIPPAGE_BPS = 50
+const MAX_SLIPPAGE_BPS = 10_000 // 100% in BPS
+const ONE_HUNDRED_PERCENT_DECIMAL = 1 // 100 in decimal percentage
+
+const SCALE = 10n ** 18n // 18 decimal places of precision. Used to avoid depending on Big Decimal libraries
```
Comment (alfetopito) [inline src/trading/suggestSlippageBpsFromFee.ts:7]: Any particular reason for using 18 decimals precision?
Is it because most tokens have 18 decimals?
How would that work with tokens that have less or more than 18 decimals?

File: src/trading/suggestSlippageBpsFromFee.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (anxolin) [reply src/trading/suggestSlippageBpsFromFee.ts:65]: Exactly!

File: package.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@cowprotocol/cow-sdk",
-  "version": "6.0.0-RC.36",
+  "version": "6.0.0-RC.38",
```
[the remark this answers is not in the document]
Comment (anxolin) [reply package.json]: But this has already the breaking changes from the BridgingSDK. I think this is what aave should use, no? I based this PR in the latest from main

File: src/trading/suggestSlippageBpsFromFee.ts (hunk 4 above)
Comment (anxolin) [reply src/trading/suggestSlippageBpsFromFee.ts:7]: No particular reason. Just to get a ton of precision I don't need. 

I added 6 decimals now, which is enough.

File: src/order-book/quoteAmountsAndCostsUtils.ts [hunk 6]
```diff
@@ -11,9 +11,40 @@ export interface QuoteAmountsAndCostsParams {
...
 const ONE_HUNDRED_BPS = BigInt(100 * 100)
 
-export function getQuoteAmountsAndCosts(params: QuoteAmountsAndCostsParams): QuoteAmountsAndCosts {
-  const { orderParams, sellDecimals, buyDecimals, slippagePercentBps } = params
-  const partnerFeeBps = params.partnerFeeBps ?? 0
+export function getQuoteAmountsWithCosts(params: {
+  sellDecimals: number
+  buyDecimals: number
+  orderParams: OrderParameters
+}) {
```
Comment (shoom3301) [inline src/order-book/quoteAmountsAndCostsUtils.ts:18]: Would be great to keep the return type `QuoteAmountsAndCosts` to be sure if we change the function implementation it still will match the type

**Comment by anxolin** [comment]
Let me know if there's new comments. I need to consolidate the PR. And sasha, i will address your comment too
